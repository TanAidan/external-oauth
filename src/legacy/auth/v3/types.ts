import { Request } from 'express'
import Integration from '../../functions/integration'
import Knex from 'knex'

export interface IAuthContext {
  clientId: string
  connectParams: any
  buid?: string
  setupId: string
  isCallback?: boolean
  store: Knex
  configuration: SetupRecord
}

interface SetupRecord {
  setup_id: string
  scopes: string[] | undefined
  buid: string
  configuration: any
}

interface IContextSessionData {
  session: {
    context: any
  }
}

export enum EDashboardApiErrorCodes {
  ENVIRONMENT_NOT_FOUND = 'ENVIRONMENT_NOT_FOUND',
  INTEGRATION_NOT_FOUND = 'INTEGRATION_NOT_FOUND'
}

export interface IDashboardApiResponse {
  errors: [
    {
      code: EDashboardApiErrorCodes
      message: string
    }
  ]
  data: {
    viewer: {
      canInvoke: boolean
      hasAccessTo: boolean
      clientId: string
      setupId: string
      buid: string
    }
  }
}

export type TConnectContextRequest = Request &
  IContextSessionData &
  IAuthContext & {
    params: {
      buid: string
    }
    query: {
      clientId: string
      setupId: string
    }
    integration?: Integration
    fullLogs?: boolean
  }

export type TCallbackContextRequest = Request & IContextSessionData & IAuthContext

export enum ELocalRedirectMethod {
  Inline = 'inline',
  Localhost = 'localhost'
}

type TNameValues = Record<string, any>

export type TIntegrationConfig = TNameValues & {
  authType: EAuthType
  config?: {
    scope?: string[]
    state?: any
  }
}

export interface IAuthConfig {
  integrationConfig: TIntegrationConfig
  setupDetails: TNameValues
}

export enum EAuthType {
  NoAuth = 'NO_AUTH',
  OAuth1 = 'OAUTH1',
  OAuth2 = 'OAUTH2'
}

interface IAuthConfigSessionData {
  session: {
    authConfig: IAuthConfig
  }
}

export type TConnectConfigRequest = Request & IAuthContext & IAuthConfig & IAuthConfigSessionData

export type TCallbackConfigRequest = Request & IAuthConfig & IAuthConfigSessionData

export interface IAuthId {
  authId: string
}

export type TAuthIdRequest = Request &
  IAuthContext &
  IAuthId & {
    session: {
      authId: string
      context: {
        setupId: string
      }
    }
  }

export interface IOAuth1Credentials {
  accessToken: string
  tokenSecret: string
  consumerKey: string
  consumerSecret: string
  expiresIn: number
}

export interface IOAuth2Credentials {
  accessToken: string
  refreshToken: string
  idToken: string
  idTokenJwt?: any
  expiresIn: number
}

export interface IWithAuthCredentials {
  credentials: IOAuth1Credentials | IOAuth2Credentials
  tokenResponse?: OAuthTokenResponse
}

export type TAuthenticateRequest = Request & IAuthContext & IAuthConfig & IWithAuthCredentials

export type AuthSuccessRequest = Request &
  IAuthContext &
  IAuthId &
  IWithAuthCredentials & { buid: string; integrationConfig?: TIntegrationConfig } & { store: Knex }

export type TRevokeRequest = Request & { buid: string }

export type TErrorHandlerRequest = Request & Partial<IAuthContext> & Partial<IAuthId>

export interface IFetchAuthDetailsParams {
  buid: string
  scopedUserDataTableName: string
  servicesTableName: string
  integration: Integration
  environmentIdentifier: string
  authId?: string
  setupId?: string
  configuration: any
  setupIdFromRequest?: boolean
  store: any
}

export interface BasicAuthDetails {
  username: string
  password: string
}

export interface ApiKeyAuthDetails {
  apiKey: string
}

export enum OAuth1SignatureMethod {
  HmacSha1 = 'HMAC-SHA1',
  RsaSha1 = 'RSA-SHA1',
  PlainText = 'PLAINTEXT'
}

interface OAuthAuthDetails {
  callbackParams: Record<string, string>
  tokenResponse?: OAuthTokenResponse
  updatedAt: number
}

export interface OAuth1AuthDetails extends OAuthAuthDetails {
  accessToken: string
  expiresIn: number
  tokenSecret: string
  consumerKey: string
  consumerSecret: string
  signatureMethod: 'HMAC-SHA1' | 'PLAINTEXT' | 'RSA-SHA1'
}

export interface OAuth2AuthDetails extends OAuthAuthDetails {
  accessToken: string
  callbackParams: Record<string, string>
  clientId: string
  clientSecret: string
  expiresIn: number
  idToken?: string
  idTokenJwt?: any
  refreshToken?: string
}

export interface OAuthTokenResponse {
  body: any
  headers: Record<string, string>
}

export type AuthDetails = BasicAuthDetails | ApiKeyAuthDetails | OAuth1AuthDetails | OAuth2AuthDetails | undefined
