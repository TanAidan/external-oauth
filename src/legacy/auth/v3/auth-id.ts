import { Response, NextFunction } from 'express'
import { TAuthIdRequest } from './types'
import { v1 as uuidv1 } from 'uuid'

export const connectAuthId = (req: TAuthIdRequest, res: Response, next: NextFunction) => {
  const currentAuthId = req.query.authId as string
  const authId = currentAuthId || uuidv1()

  req.authId = authId
  req.session.authId = authId

  next()
}

export const callbackAuthId = (req: TAuthIdRequest, _res: Response, next: NextFunction) => {
  const { authId } = req.session
  req.authId = authId

  next()
}
