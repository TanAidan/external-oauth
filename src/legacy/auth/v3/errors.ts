import { UserError } from '../../errors'
import { EAuthType } from './types'
import { Request } from 'express'
import { GrantType } from './strategies/oauth2'

export class NoAuthInProgress extends UserError {
  constructor() {
    super(
      `The callback was invoked whilst no authentication was in progress. ` +
        `Either too much time has elapsed between starting authentication and the callback being invoked, ` +
        `or the callback was invoked without first calling the integration's authentication endpoint. Please try again`,
      422,
      'NO_AUTH_IN_PROGRESS'
    )
  }
}
export class InvalidAuthType extends UserError {
  constructor(authType: string) {
    const allowedValues = Object.values(EAuthType)
      .map(authType => `'${authType}'`)
      .join(', ')
    super(
      `Invalid authType '${authType}'. Please check the integration's auth.config.json file. ` +
        `Allowed values are ${allowedValues}`,
      422,
      'INVALID_AUTH_TYPE'
    )
  }
}

export class InvalidGrantType extends UserError {
  constructor(grantType: string) {
    const allowedValues = Object.values(GrantType)
      .map(grantType => `'${grantType}'`)
      .join(', ')
    super(
      `Invalid grantType '${grantType}'. Please check the API's configuration. ` +
        `Allowed values are ${allowedValues}`,
      422,
      'INVALID_GRANT_TYPE'
    )
  }
}

export class OAuthOnlyEndpoint extends UserError {
  constructor(authType: string) {
    super(
      "This endpoint is only for 'OAUTH1' and 'OAUTH2' auth types " +
        `but the integration is configured with authType '${authType}'`,
      422,
      'OAUTH_ONLY_ENDPOINT'
    )
  }
}

export class AuthenticationFailed extends UserError {
  readonly error: string
  readonly errorDescription?: string

  static fromOAuthRequest(req: Request, info: any) {
    if (req.query.error) {
      return new AuthenticationFailed(req.query)
    }

    return new AuthenticationFailed(info)
  }

  constructor(details: any) {
    super(`Authentication failed:\n${JSON.stringify(details, undefined, 2)}`, 403, 'AUTHENTICATION_FAILED')

    if (details.error) {
      this.error = details.error
      this.errorDescription = details.error_description
    } else {
      this.errorDescription = this.error = JSON.stringify(details)
    }
  }
}

export class AccessTokenExpired extends UserError {
  constructor() {
    super(
      `The access token associated with your Auth Id has expired and token refreshing is not supported for \
this API. Please reconnect to obtain a new access token.

Consult the API provider's documentation to determine whether refresh tokens can be enabled for this API`,
      403,
      'ACCESS_TOKEN_EXPIRED'
    )
  }
}

export class InconsistentSetupId extends UserError {
  constructor() {
    super(
      'The provided setupId does not match the setupId used when connecting. ' +
        'Please reconnect using this setupId, or retry this call with the ' +
        'setupId corresponding to the authId',
      422,
      'INCONSISTENT_SETUP_ID'
    )
  }
}

export class InvalidAuthId extends UserError {
  constructor(buid: string, authId: string) {
    super(
      `No auth details were found using authId '${authId}' for API '${buid}'

Please try again with a valid authId or connect with this authId`,
      404,
      'INVALID_AUTH_ID'
    )
  }
}

export class MissingAuthId extends UserError {
  constructor(buid: string) {
    super(
      `You must supply an authId to use the '${buid}' API

Please try again with a valid authId.

See the following link for information on how to obtain an authId: https://docs.bearer.sh/faq/connect-button`,
      401,
      'MISSING_AUTH_ID'
    )
  }
}

export class CredentialsNotConfigured extends UserError {
  constructor(buid: string) {
    super(
      `Missing credentials for the '${buid}' API

Please configure the credentials in the dashboard and try again.

See the following link for information: https://docs.bearer.sh/dashboard/apis`,
      422,
      'CREDENTIALS_NOT_CONFIGURED'
    )
  }
}
