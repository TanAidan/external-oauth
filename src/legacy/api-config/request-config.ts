import merge from 'merge'
import { OAuth } from 'oauth'

import { expandTemplate, UndefinedVariable } from './template'
import { EAuthType, AuthDetails, OAuth1AuthDetails } from '../auth/v3/types'
import { MissingApiConfigHeader, InvalidApiConfig, MissingApiConfigConnectParam } from './errors'

const labels: Record<keyof RequestConfig, string> = {
  baseURL: 'Request Base URL',
  headers: 'Request Header',
  params: 'Request Param'
}

export const expandRequestConfig = (params: ExpandRequestConfigParams): RequestConfig => {
  try {
    return expandTemplate(params.requestConfig, getVariables(params))
  } catch (e) {
    if (!(e instanceof UndefinedVariable)) {
      throw e
    }

    const { templatePath, variableName } = e
    const [attributeName, subAttributeName] = templatePath.split('.')
    const label = labels[attributeName]

    if (e.variableName.startsWith('connectParams.')) {
      throw new MissingApiConfigConnectParam(label, subAttributeName, variableName)
    }

    if (e.variableName.startsWith('headers.')) {
      throw new MissingApiConfigHeader(label, subAttributeName, variableName)
    }

    throw new InvalidApiConfig(label, subAttributeName, variableName)
  }
}

const tryExpandBaseURL = (baseURL: string, variables: any) => {
  try {
    return expandTemplate(baseURL, variables)
  } catch (e) {
    return baseURL
  }
}

const getVariables = ({
  auth,
  authType,
  connectParams,
  headers,
  method,
  path,
  requestConfig
}: ExpandRequestConfigParams) => {
  const partialVariables = { auth, connectParams, headers }
  // We want any expansion errors to be caught later in the process when
  // we have the proper context for error messages
  const baseURL = tryExpandBaseURL(requestConfig.baseURL, partialVariables)

  return merge.recursive(true, partialVariables, {
    auth: additionalAuthVariables({ auth, authType, baseURL, method, path })
  })
}

const additionalAuthVariables = ({ auth, authType, baseURL, method, path }: AdditionalAuthVariablesParams) => {
  switch (authType) {
    case EAuthType.OAuth1:
      return { oauth1: getOAuth1Credentials({ baseURL, method, path, auth: auth as OAuth1AuthDetails }) }
    default:
      return {}
  }
}

export const getOAuth1Credentials = ({ baseURL, method, path, auth }: IGetOAuth1CredentialsParams) => {
  const { consumerKey, consumerSecret, accessToken, tokenSecret, signatureMethod } = auth

  const absUrl = `${baseURL.replace(/\/+$/, '')}/${path.replace(/^\/+/, '')}`

  const oauth = new OAuth(
    '', // requestUrl
    '', // accessUrl
    consumerKey,
    consumerSecret,
    '1.0', // version
    '', // authorize_callback
    signatureMethod
  )

  return oauth.authHeader(absUrl, accessToken, tokenSecret, method).replace(/^OAuth /, '')
}

interface ExpandRequestConfigParams {
  auth: AuthDetails
  authType: EAuthType
  connectParams: Record<string, string>
  headers: Record<string, string>
  method: string
  path: string
  requestConfig: RequestConfig
}

export interface RequestConfig {
  baseURL: string
  headers: Record<string, string>
  params: Record<string, string>
}

interface AdditionalAuthVariablesParams {
  auth: AuthDetails
  authType: EAuthType
  baseURL: string
  method: string
  path: string
}

interface IGetOAuth1CredentialsParams {
  auth: OAuth1AuthDetails
  baseURL: string
  method: string
  path: string
}
