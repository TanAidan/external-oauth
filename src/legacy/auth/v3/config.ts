import { asyncMiddleware } from '../../errorHandler'
import { IAuthConfig, EAuthType, TConnectConfigRequest, TCallbackConfigRequest } from './types'
import { NextFunction, Response } from 'express'
import {
  getConfig
  // getSetupDetails
} from '../clients/integrations'
import { InvalidAuthType } from './errors'
import { expandAuthConfig } from '../../api-config/auth-config'

const getAuthConfig = async (req: TConnectConfigRequest) => {
  const { connectParams, buid, configuration } = req

  return {
    setupDetails: configuration,
    integrationConfig: expandAuthConfig({
      connectParams,
      authConfig: (await getConfig({
        buid: buid!
      })) as any
    })
  } as IAuthConfig
}

const copyConfig = (src: IAuthConfig, dest: IAuthConfig) => {
  dest.integrationConfig = src.integrationConfig
  dest.setupDetails = src.setupDetails
}

export const connectConfig = asyncMiddleware(async (req: TConnectConfigRequest, _res: Response, next: NextFunction) => {
  const authConfig = await getAuthConfig(req)
  const { authType } = authConfig.integrationConfig

  if (!Object.values(EAuthType).includes(authType)) {
    throw new InvalidAuthType(authType)
  }

  copyConfig(authConfig, req)
  req.session.authConfig = authConfig

  next()
})

export const callbackConfig = (req: TCallbackConfigRequest, _res: Response, next: NextFunction) => {
  copyConfig(req.session.authConfig, req)

  next()
}
