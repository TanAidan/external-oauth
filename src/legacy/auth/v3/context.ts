import { TConnectContextRequest, TCallbackContextRequest } from './types'
import { NextFunction, Response } from 'express'
import { UserError } from '../../errors'
import { NoAuthInProgress } from './errors'
import { asyncMiddleware } from '../../errorHandler'

const getConnectParams = (params?: any) => {
  if (params === undefined) {
    return {}
  }

  if (typeof params !== 'object' || Array.isArray(params)) {
    throw new InvalidConnectParams()
  }

  for (const [name, value] of Object.entries(params)) {
    if (typeof value !== 'string' || !/^[\w\s.-]*$/.test(value)) {
      throw new InvalidConnectParam(name)
    }
  }

  return params
}

export const connectContext = (req: TConnectContextRequest, res: Response, next: NextFunction) => {
  const { buid, setupId } = req

  const connectParams = getConnectParams(req.query.params)

  req.session.context = {
    connectParams,
    setupId,
    buid
  }

  req.connectParams = connectParams

  next()
}

export const callbackContext = asyncMiddleware(
  async (req: TCallbackContextRequest, _res: Response, next: NextFunction) => {
    if (!req.session.context) {
      throw new NoAuthInProgress()
    }

    const { buid, connectParams } = req.session.context

    req.isCallback = true
    req.connectParams = connectParams
    req.buid = buid
    req.setupId = req.session.authConfig.setupDetails.setup_id

    next()
  }
)

class InvalidConnectParams extends UserError {
  constructor() {
    super(
      // tslint:disable:max-line-length
      `Incorrect format for connect parameters'

Connect parameters must be sent as query parameters of the form \`params[name]=value\` eg. \`params[subdomain]=my-app\`'`,
      // tslint:enable:max-line-length
      400,
      'INVALID_CONNECT_PARAMS'
    )
  }
}

class InvalidConnectParam extends UserError {
  constructor(name: string) {
    super(
      `Incorrect format for connect parameter '${name}'

Connect parameters may contain alphanumeric and space characters, or any of the following symbols '_-.'

Refer to this link for further information: https://docs.bearer.sh/faq/connect-button`,
      400,
      'INVALID_CONNECT_PARAM'
    )
  }
}
