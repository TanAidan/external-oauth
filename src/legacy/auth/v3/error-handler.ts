import { TErrorHandlerRequest } from './types'
import { Response, NextFunction } from 'express'
import { AuthenticationFailed, NoAuthInProgress } from './errors'
import { isEmpty } from '../../functions/utils'

const respondWithOAuthError = (
  res: Response,
  { statusCode, code, message }: { statusCode: number; code: string; message: string }
) => {
  res.header('Content-Type', 'text/html')
  res.status(statusCode)
  res.render('auth/error', { error: `${code}: ${message}` })
}

const respondWithCallbackError = (req: TErrorHandlerRequest, res: Response, err: AuthenticationFailed) => {
  res.header('Content-Type', 'text/html')
  res.status(err.statusCode)
  res.render('auth/callback', {
    authId: req.authId,
    error: err.error,
    error_description: err.errorDescription,
    integrationUuid: req.buid
  })
}

const respondWithCallbackPlaceholder = (
  req: TErrorHandlerRequest,
  res: Response,
  { statusCode, code, message }: { statusCode: number; code: string; message: string }
) => {
  res.header('Content-Type', 'text/html')
  res.status(statusCode)
  res.render('auth/placeholder', {
    error: `${code}: ${message}`,
    callbackUrl: req.originalUrl,
    method: req.method
  })
}

export const errorHandler = (err: any, req: TErrorHandlerRequest, res: Response, next: NextFunction) => {
  if (res.headersSent) {
    return next(err)
  }

  if (err.statusCode) {
    if (err instanceof AuthenticationFailed) {
      return respondWithCallbackError(req, res, err)
    }

    if (err instanceof NoAuthInProgress && isEmpty(req.query)) {
      return respondWithCallbackPlaceholder(req, res, err)
    }

    return respondWithOAuthError(res, err)
  }
  respondWithOAuthError(res, {
    statusCode: 500,
    code: 'INTERNAL_ERROR',
    message: 'Encountered an unexpected error. Please contact support'
  })
}
