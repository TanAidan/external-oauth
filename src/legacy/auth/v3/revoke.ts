import { asyncMiddleware } from '../../errorHandler'
import { TRevokeRequest } from './types'
// import { revokeAuthV3 } from '../../clients/integrations'
import { Response } from 'express'
import { MissingParameter } from '../../errors'

export const revoke = asyncMiddleware(async (req: TRevokeRequest, res: Response) => {
  // const { authId } = req.params
  // const { buid } = req
  const { clientId } = req.query

  if (!clientId) {
    throw new MissingParameter('clientId')
  }

  // await revokeAuthV3({
  //   clientId,
  //   buid,
  //   authId,
  //   servicesTableName: req.stageVariables.servicesTableId
  // })

  res.status(200).end()
})
