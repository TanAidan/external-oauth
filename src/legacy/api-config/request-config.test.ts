import { OAuth } from 'oauth'

import { expandRequestConfig, RequestConfig } from './request-config'
import { EAuthType, OAuth1SignatureMethod, OAuth1AuthDetails, AuthDetails, BasicAuthDetails } from '../auth/v3/types'

jest.mock('oauth')

describe('expandRequestConfig', () => {
  const headers = {
    authorization: 'Bearer ${ auth.password }',
    h2: 'Some ${headers.some_value}'
  }

  const connectParams = {
    conParam: 'test-con-param'
  }

  const params = {
    user: '${auth.username}',
    p2: '${  headers.other_value  }',
    p3: '${connectParams.conParam}'
  }

  const defaultRequestConfig: RequestConfig = {
    headers,
    params,
    baseURL: 'https://${headers.instance}.example.com/${auth.username}'
  }

  const basicAuth = { username: 'test-user', password: 'test-pass' }
  const requestHeaders = { some_value: 'test-some-value', other_value: 'test-other-value', instance: 'test-instance' }

  const setup = ({
    auth = basicAuth,
    authType = EAuthType.NoAuth,
    requestConfig = defaultRequestConfig
  }: { auth?: AuthDetails; authType?: EAuthType; requestConfig?: RequestConfig } = {}) => ({
    auth,
    authType,
    connectParams,
    requestConfig,
    headers: requestHeaders,
    method: 'GET',
    path: '/super-secure'
  })

  it('fills in the variables', () => {
    const params = setup()

    expect(expandRequestConfig(params)).toMatchSnapshot()
  })

  describe('baseURL', () => {
    it('returns an error if it contains an invalid variable', () => {
      const params = setup({
        requestConfig: { ...defaultRequestConfig, baseURL: 'http://${invalid}.example.com' }
      })

      expect(() => expandRequestConfig(params)).toThrowErrorMatchingSnapshot()
    })

    it('returns an error if a required header is missing', () => {
      const params = setup({
        requestConfig: { ...defaultRequestConfig, baseURL: 'http://${headers.missing}.example.com' }
      })

      expect(() => expandRequestConfig(params)).toThrowErrorMatchingSnapshot()
    })
  })

  describe('headers', () => {
    it('returns an error if one contains an invalid variable', () => {
      const params = setup({
        requestConfig: { ...defaultRequestConfig, headers: { test: 'Bearer ${invalid}' } }
      })

      expect(() => expandRequestConfig(params)).toThrowErrorMatchingSnapshot()
    })

    it('returns an error if a required header is missing', () => {
      const params = setup({
        requestConfig: { ...defaultRequestConfig, headers: { test: 'Bearer ${headers.missing}' } }
      })

      expect(() => expandRequestConfig(params)).toThrowErrorMatchingSnapshot()
    })

    it('returns an error if a required connect param is missing', () => {
      const params = setup({
        requestConfig: { ...defaultRequestConfig, headers: { test: 'Bearer ${connectParams.missing}' } }
      })

      expect(() => expandRequestConfig(params)).toThrowErrorMatchingSnapshot()
    })
  })

  describe('params', () => {
    it('returns an error if one contains an invalid variable', () => {
      const params = setup({
        requestConfig: { ...defaultRequestConfig, params: { test: 'Bearer ${invalid}' } }
      })

      expect(() => expandRequestConfig(params)).toThrowErrorMatchingSnapshot()
    })

    it('returns an error if a required header is missing', () => {
      const params = setup({
        requestConfig: { ...defaultRequestConfig, params: { test: 'Bearer ${headers.missing}' } }
      })

      expect(() => expandRequestConfig(params)).toThrowErrorMatchingSnapshot()
    })

    it('returns an error if a required connect param is missing', () => {
      const params = setup({
        requestConfig: { ...defaultRequestConfig, params: { test: 'Bearer ${connectParams.missing}' } }
      })

      expect(() => expandRequestConfig(params)).toThrowErrorMatchingSnapshot()
    })
  })

  describe('when authType is OAuth1', () => {
    const oauth1Creds = 'test-oauth1-creds'
    // const authHeader = jest.fn(() => `OAuth ${oauth1Creds}`)

    const setupOAuth1 = ({ baseURL = 'https://api.example.com' }: { baseURL?: string } = {}) => {
      const requestConfig = {
        baseURL,
        headers: {
          test: '${auth.oauth1}'
        },
        params: {}
      }

      return setup({
        requestConfig,
        auth: {
          signatureMethod: OAuth1SignatureMethod.HmacSha1,
          consumerKey: 'test-consumer-key',
          consumerSecret: 'test-consumer-secret',
          accessToken: 'test-access-token',
          tokenSecret: 'test-token-secret'
        } as OAuth1AuthDetails,
        authType: EAuthType.OAuth1
      })
    }

    beforeEach(() => {
      // mocked(OAuth)
      //   .mockImplementationOnce(() => ({ authHeader }))
      //   .mockClear()
      // authHeader.mockClear()
    })

    it('sets `auth.oauth1` to the encoded credentials for the auth header', async () => {
      const params = setupOAuth1()

      expect(expandRequestConfig(params).headers.test).toBe(oauth1Creds)
      expect(OAuth).toMatchSnapshot()
    })

    it('forms the correct URL when the baseURL ends in a slash', async () => {
      const params = setupOAuth1({ baseURL: 'https://api.example.com/path/' })

      expandRequestConfig(params)

      // expect(authHeader.mock.calls[0][0]).toBe('https://api.example.com/path/super-secure')
    })
  })

  describe('when authType is Basic', () => {
    const setupBasic = (usePassword: boolean) => {
      const requestConfig = {
        baseURL: '',
        headers: {
          test: '${auth.basic}'
        },
        params: {}
      }

      return setup({
        requestConfig,
        auth: {
          username: 'test-user',
          password: usePassword ? 'test-pass' : undefined
        } as BasicAuthDetails,
        authType: EAuthType.NoAuth
      })
    }

    it('sets `auth.basic` to the encoded credentials for the auth header', async () => {
      const params = setupBasic(true)

      expect(expandRequestConfig(params).headers.test).toBe('dGVzdC11c2VyOnRlc3QtcGFzcw==')
    })

    it('supports empty passwords', async () => {
      const params = setupBasic(false)

      expect(expandRequestConfig(params).headers.test).toBe('dGVzdC11c2VyOg==')
    })
  })
})
