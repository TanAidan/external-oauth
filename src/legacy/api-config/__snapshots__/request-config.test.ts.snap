// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expandRequestConfig baseURL returns an error if a required header is missing 1`] = `"The API's Request Base URL configuration requires the 'missing' header to be passed with each request."`;

exports[`expandRequestConfig baseURL returns an error if it contains an invalid variable 1`] = `"The API's Request Base URL has been configured with an invalid variable \${invalid}."`;

exports[`expandRequestConfig fills in the variables 1`] = `
Object {
  "baseURL": "https://test-instance.example.com/test-user",
  "headers": Object {
    "authorization": "Bearer test-pass",
    "h2": "Some test-some-value",
  },
  "params": Object {
    "p2": "test-other-value",
    "p3": "test-con-param",
    "user": "test-user",
  },
}
`;

exports[`expandRequestConfig headers returns an error if a required connect param is missing 1`] = `"The API's Request Header 'test' configuration requires the 'missing' query parameter to be passed when connecting."`;

exports[`expandRequestConfig headers returns an error if a required header is missing 1`] = `"The API's Request Header 'test' configuration requires the 'missing' header to be passed with each request."`;

exports[`expandRequestConfig headers returns an error if one contains an invalid variable 1`] = `"The API's Request Header 'test' has been configured with an invalid variable \${invalid}."`;

exports[`expandRequestConfig params returns an error if a required connect param is missing 1`] = `"The API's Request Param 'test' configuration requires the 'missing' query parameter to be passed when connecting."`;

exports[`expandRequestConfig params returns an error if a required header is missing 1`] = `"The API's Request Param 'test' configuration requires the 'missing' header to be passed with each request."`;

exports[`expandRequestConfig params returns an error if one contains an invalid variable 1`] = `"The API's Request Param 'test' has been configured with an invalid variable \${invalid}."`;

exports[`expandRequestConfig when authType is OAuth1 sets \`auth.oauth1\` to the encoded credentials for the auth header 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "",
      "",
      "test-consumer-key",
      "test-consumer-secret",
      "1.0",
      "",
      "HMAC-SHA1",
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Object {
        "authHeader": [MockFunction] {
          "calls": Array [
            Array [
              "https://api.example.com/super-secure",
              "test-access-token",
              "test-token-secret",
              "GET",
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": "OAuth test-oauth1-creds",
            },
          ],
        },
      },
    },
  ],
}
`;
